File: 0875-koko-eating-bananas/0875-koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        return minEating(piles,h);
        
    }

    private static boolean canEatAll(int[] piles,int h,int k){
        int hours = 0;
        for(int pile:piles){
            hours += (pile + k - 1) / k;

            if(hours>h) return false;
        }

        return hours<=h;
    }

    public static int minEating(int[] piles,int h){
        int low=0;
        int high=0;
        for(int pile:piles){
            high=Math.max(high,pile);
        }
        int ans=high;

        while(low<=high){
            int mid = (low+high)/2;
            if(canEatAll(piles,h,mid)){
                ans=mid;
                high=mid-1;
            }else{
                low=mid+1;
            }
        }
        return ans;
    }

}

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int low=1;
        int high=0;
        for (int pile:piles){
            high=max(high,pile);
        }
        while(low<=high){
            int mid=low+(high-low)/2;
            if(canEat(piles,mid,h)){
                high=mid-1;
            }else{
                low=mid+1;
            }
        }
        return low;
    }
    bool canEat(vector<int>& piles,int speed,int h){
        int time=0;
        for(int pile:piles){
            time+=(pile+speed-1)/speed;
            if(time>h) return false;
        }

        return true;
    }
};